### PR TITLE
update to 2018 edition

### PR DESCRIPTION
**Pull request by benmkw**
I made some mechanic changes to update to 2018 edition but don't have strong opinions about them so feel free to request changes and it would also maybe good to do a benchmark comparison because I actually don't know if once_cell vs. lazy_static might regress here as one example.

updating nalgebra turned out to be harder than expected :/ 

I did not yet open the bbox PR as I want to wait for your feedback, but applied the same things as here in the linked patch at the botton of the Cargo.toml.

### CODE REVIEW

**Review by hmeyer — DISMISSED** [review]

File: Cargo.toml [hunk 1]
```diff
@@ -16,11 +16,10 @@ path = "src/lib.rs"
 
 [dependencies]
 alga = "0.9"
-nalgebra = "0.19"
+nalgebra = "0.22.1"
 rand = "0.7"
-time = "0.1"
 rayon = "1.2"
-lazy_static = "1.4"
+once_cell = "1.4.1"
```
Comment (hmeyer) [inline Cargo.toml]: What's the rationale for changing from lazy_static to once_cell?
Can we do this in a separate PR?

File: src/manifold_dual_contouring.rs [hunk 2]
```diff
@@ -391,18 +395,20 @@ fn subsample_octtree<S: RealField + Float + From<f32>>(base: &[Vertex<S>]) -> Ve
...
 impl Timer {
     fn new() -> Timer {
-        Timer { t: ::time::now() }
+        Timer {
+            t: std::time::Instant::now(),
+        }
     }
-    fn elapsed(&mut self) -> ::time::Duration {
-        let now = ::time::now();
+    fn elapsed_millis(&mut self) -> u128 {
```
Comment (hmeyer) [inline src/manifold_dual_contouring.rs]: can we change this to:
fn elapsed(&mut self) -> std::time::Duration 
?

File: Cargo.toml [hunk 3]
```diff
@@ -36,3 +36,6 @@ harness = false
 [badges]
 travis-ci = { repository = "hmeyer/tessellation", branch = "master" }
 codecov = { repository = "hmeyer/tessellation", branch = "master", service = "github" }
+
+[patch.crates-io]
+bbox = { git = "https://github.com/benmkw/bbox" }
```
Comment (hmeyer) [inline Cargo.toml]: I just cherry-picked your 6ae2799.
So I guess we can just dep on bbox = "0.10" ?

File: Cargo.toml (hunk 1 above)
Comment (benmkw) [reply Cargo.toml]: "once_cell also has a Lazy<T> type, build on top of OnceCell which provides the same API as the lazy_static! macro, but without using any macro" and "The API of once_cell is being proposed for inclusion in std" from https://github.com/matklad/once_cell 

could definitely be a different PR

File: src/manifold_dual_contouring.rs (hunk 2 above)
Comment (benmkw) [reply src/manifold_dual_contouring.rs]: you have to do .to_millis() or whatever at every callside but I agree that the return type would read better, idk

File: Cargo.toml (hunk 3 above)
Comment (benmkw) [reply Cargo.toml]: cool, thanks, sure!

I could use some help with updating nalgebra if you have time, it seems they are in disagreement about their traits a bit, maybe I should write up the problem more accurately but I basically tried to resolve it for two hours and failed :D

Comment (benmkw) [reply Cargo.toml]: seems like we need to bump it in implicit3d as well, should I do the same changes there as well? sorry these changes spread so much :/

File: Cargo.toml (hunk 1 above)
Comment (hmeyer) [reply Cargo.toml]: Looks like a reasonable change to me.
But let's move this to a separate PR, please.

File: src/manifold_dual_contouring.rs (hunk 2 above)
Comment (hmeyer) [reply src/manifold_dual_contouring.rs]: I think we could just use the Debug trait of Duration, and change the println!(".... {:}...", t.elapsed()); to
println!(".... {:?}...", t.elapsed());

File: Cargo.toml (hunk 3 above)
Comment (hmeyer) [reply Cargo.toml]: Yes, updating nalgebra is a pain.
I usually add a minor version bump for all libs when updating nalgebra.

File: src/manifold_dual_contouring.rs (hunk 2 above)
Comment (benmkw) [reply src/manifold_dual_contouring.rs]: sure that works as well

**Comment by benmkw** [comment]
I think I addressed the review comments and rebased.

File: Cargo.toml (hunk 3 above)
Comment (hmeyer) [reply Cargo.toml]: I updated bbox, tessellation and implicit3d to nalgebra 0.22.

Comment (benmkw) [reply Cargo.toml]: yeah but it still does not work for tesselation :D at least I think it may be easier for you to apply the changes but I can look into it as well if you prefer just need some more time :)

Comment (benmkw) [reply Cargo.toml]: ok so i got it compiling but sometimes there are now 3 different RealField requirements which seems a bit odd, maybe I can remove some and then its fine again

Comment (benmkw) [reply Cargo.toml]: it seems a bit odd that tesselation now depends on traits from implicit3d rather than alga and nalgebra as this defeats the purpose of generics a bit as one now has to depend on implciit3d just to get the traits if I understand correctly. 

Maybe it would thus be better to use the actual two traits even though thats noisier.

**Comment by hmeyer** [comment]
Nice! Thank you very much!
